fe/fe-core/src/main/java/com/starrocks/qe/scheduler/dag/FragmentSequence.java

```diff
@@ -0,0 +1,146 @@
+// │   ┌────┴───┐       │        │
+// │   │        │       │        │
+// │   ▼        ▼       │        │
+// │ Exchange  Exchange │        │
+// └───┬────────┬───────┼────────┘
+//     │        │       │
+//     │        │       │
+//     │        │       │
+//     ▼        ▼       ▼
+//     F1       F2      F3
+// the fragment child schedule order should be [F1, F2, F3]
+// In phased schedule we should schedule F3 firstly. When F3 finished, then schedule F2
+public class FragmentSequence {
+
+    private final ExecutionDAG dag;
+    private final List<PlanFragmentId> scheduleOrderList;
+
+    public FragmentSequence(ExecutionDAG dag, List<PlanFragmentId> fragmentIds) {
+        this.dag = dag;
+        this.scheduleOrderList = fragmentIds;
+    }
+
+    public ExecutionFragment getAt(int i) {
+        final PlanFragmentId fragmentId = scheduleOrderList.get(i);
+        return dag.getFragment(fragmentId);
+    }
+
+    public static class Builder {
+        private static void buildExgNodeIdToSourceFragmentId(ExecutionFragment fragment,
+                                                             Map<PlanNodeId, PlanFragmentId> exchangeToSource) {
+            final DataSink sink = fragment.getPlanFragment().getSink();
+            if (sink instanceof DataStreamSink) {
+                exchangeToSource.put(sink.getExchNodeId(), fragment.getFragmentId());
+            } else if (sink instanceof MultiCastDataSink) {
+                for (DataStreamSink dataStreamSink : ((MultiCastDataSink) sink).getDataStreamSinks()) {
+                    exchangeToSource.put(dataStreamSink.getExchNodeId(), fragment.getFragmentId());
+                }
+            }
+            for (int i = 0; i < fragment.childrenSize(); i++) {
+                buildExgNodeIdToSourceFragmentId(fragment.getChild(i), exchangeToSource);
+            }
+        }
+
+        public static Map<PlanFragmentId, FragmentSequence> build(ExecutionDAG dag, ExecutionFragment root) {
+
+            Map<PlanNodeId, PlanFragmentId> exchangeToSource = Maps.newHashMap();
+            buildExgNodeIdToSourceFragmentId(root, exchangeToSource);
+
+            Map<PlanFragmentId, FragmentSequence> result = Maps.newHashMap();
+            buildIfNotExists(dag, root, exchangeToSource, result);
+
+            return result;
+        }
+
+        private static void buildIfNotExists(ExecutionDAG dag, ExecutionFragment node,
+                                             Map<PlanNodeId, PlanFragmentId> exchangeToFragment,
+                                             Map<PlanFragmentId, FragmentSequence> sequenceMap) {
+            if (sequenceMap.containsKey(node.getFragmentId())) {
+                return;
+            }
+
+            final List<PlanFragmentId> childFragments = Lists.newArrayList();
+            build(node.getPlanFragment().getPlanRoot(), childFragments, exchangeToFragment);
+            final FragmentSequence sequence = new FragmentSequence(dag, Lists.reverse(childFragments));
```

Comment:
After `build` and `Lists.reverse`, does the sequence contain `F1, F2, F3` not `F3, F2, F1`?

```
// ┌─────────────────────────────┐
// │                             │
// │           Join              │
// │             │               │
// │             │               │
// │        ┌────┴──────┐        │
// │        │           │        │
// │        ▼           ▼        │
// │       Join      Exchange    │
// │        │           │        │
// │        │           │        │
// │   ┌────┴───┐       │        │
// │   │        │       │        │
// │   ▼        ▼       │        │
// │ Exchange  Exchange │        │
// └───┬────────┬───────┼────────┘
//     │        │       │
//     │        │       │
//     │        │       │
//     ▼        ▼       ▼
//     F1       F2      F3
```

Reply:
Yes, what we offer here is reverse order.